docs/dotnet/dotnet-http-metrics.md

```diff
@@ -0,0 +1,383 @@
+
+`http.connection.state` MUST be one of the following:
+
+| Value  | Description |
+|---|---|
+| `active` | active state. |
+| `idle` | idle state. |
+<!-- endsemconv -->
+
+### Metric: `http.client.connection.duration`
+
+<!-- semconv metric.dotnet.http.client.connection.duration(metric_table) -->
+| Name     | Instrument Type | Unit (UCUM) | Description    |
+| -------- | --------------- | ----------- | -------------- |
+| `http.client.connection.duration` | Histogram | `s` | The duration of outbound HTTP connections. [1] |
+
+**[1]:** Meter name is `System.Net.Http`.
+<!-- endsemconv -->
+
+<!-- semconv metric.dotnet.http.client.connection.duration(full) -->
+| Attribute  | Type | Description  | Examples  | Requirement Level |
+|---|---|---|---|---|
+| `http.error.reason` | string | "Low-cardinality error reason: `timeout`, `cancellation` or one of [HTTP Request errors](https://github.com/dotnet/runtime/blob/c430570a01c103bc7f117be573f37d8ce8a129b8/src/libraries/System.Net.Http/src/System/Net/Http/HttpRequestError.cs)" [1] | `timeout`; `name_resolution_error`; `secure_connection_error` | Conditionally Required: If request has failed. |
+| [`network.protocol.version`](../general/attributes.md) | string | Version of the application layer protocol used. See note below. [2] | `1.1`; `2.0` | Recommended |
+| [`server.address`](../general/attributes.md) | string | Logical server hostname, matches server FQDN if available, and IP or socket address if FQDN is not known. | `example.com` | Recommended |
+| [`server.port`](../general/attributes.md) | int | Logical server port number | `80`; `8080`; `443` | Recommended |
+| [`server.socket.address`](../general/attributes.md) | string | Physical server IP address or Unix socket address. If set from the client, should simply use the socket's peer address, and not attempt to find any actual server IP (i.e., if set from client, this may represent some proxy server instead of the logical server). | `10.5.3.2` | Recommended: If different than `server.address`. |
+| [`url.scheme`](../url/url.md) | string | The [URI scheme](https://www.rfc-editor.org/rfc/rfc3986#section-3.1) component identifying the used protocol. | `https`; `ftp`; `telnet` | Recommended |
+
+**[1]:** **TODO: there is a [open discussion/proposal in otel for http error code](https://github.com/open-telemetry/semantic-conventions/pull/205)** Errors should be reported in snake_case.
+
+**[2]:** `network.protocol.version` refers to the version of the protocol used and might be different from the protocol client's version. If the HTTP client used has a version of `0.27.2`, but sends HTTP version `1.1`, this attribute should be set to `1.1`.
+
+`http.error.reason` has the following list of well-known values. If one of them applies, then the respective value MUST be used, otherwise a custom value MAY be used.
+
+| Value  | Description |
+|---|---|
+| `` | No error. Default value. |
```

Comment:
This contradicts the attribute definition in `http.client.duration`, which says:

> Conditionally Required: If request has failed.

Should this be empty or absent when there was no error? I remember arguments for both approach from from previous reviews, but now it's not obvious which approach did we decide to take. /cc @noahfalk

Reply:
what's the contradiction? 

empty is the same as the absence of the attribute. I.e. when there is no attribute, default value applies (which means no error).

Reply:
> empty is the same as the absence of the attribute.

I don't think those are equivalent.  If the `http.error.reason` tag absent, it means it is missing from the attribute list we emit in our [`histogram.Record()`](https://opentelemetry.io/docs/specs/otel/metrics/api/#record) call. I think this is an important behavioral difference.

Or am I missing something? What do you mean by saying `when there is no attribute, default value applies`? Is there some mechanism in the tooling that applies the default values after recording?

Reply:
I mean that here we're telling observability backends, dashboard authors, and end users that metric point without an attribute and with an empty one are the same (e.g. in [Prometheus you can filter using empty string](https://medium.com/pareture/query-results-where-label-is-not-present-in-prometheus-e1176320575d) for time series which don't have an attribute set).

We're also saying that "" means no error. 

We have other cases like this, for example, the default `network.protocol.name` is `http` (for HTTP, while `quic` is another possible value), but passing it on every measurement is not necessary. Same here, lack of error is the most frequent case, so let's optimize and pass nothing when it happens. 

Let me know how I can phrase it better or if you still see an issue.